File: runtimes/python/willyhorizont/runtime/xl.py
def escape_string(s):
    if s is None:
        return ""
    r = str(s)
    r = r.replace("\\", "\\\\")
    r = r.replace("\"", "\\\"")
    r = r.replace("\n", "\\n")
    r = r.replace("\r", "\\r")
    r = r.replace("\t", "\\t")
    return r


def json_stringify(a, pretty=False):
    p = pretty
    t = " " * 4
    s = [{"t": "v", "v": a, "d": 0}]
    r = ""
    while len(s) > 0:
        c = s.pop()
        if c["t"] == "r":
            r+= c["v"]
            continue
        v = c["v"]
        cur_d = c["d"]
        if v is None:
            r += "null"
            continue
        if isinstance(v, bool):
            r += "true" if v else "false"
            continue
        if isinstance(v, str):
            r += f"\"{escape_string(v)}\""
            continue
        if isinstance(v, (int, float)):
            r += str(v)
            continue
        if callable(v):
            r += "\"[object Function]\""
            continue
        if isinstance(v, list):
            if len(v) == 0:
                r += "[]"
                continue
            child_d = cur_d + 1
            s.append({
                "t": "r",
                "v": f"\n{t * cur_d}]" if p else "]",
                "d": cur_d
            })
            for i in range(len(v) - 1, -1, -1):
                s.append({
                    "t": "v",
                    "v": v[i],
                    "d": child_d
                })
                if i > 0:
                    s.append({
                        "t": "r",
                        "v": f",\n{t * child_d}" if p else ",",
                        "d": child_d
                    })
            s.append({
                "t": "r",
                "v": f"[\n{t * child_d}" if p else "[",
                "d": child_d
            })
            continue
        if isinstance(v, dict):
            dpl = list(v.items())
            if len(dpl) == 0:
                r += "{}"
                continue
            child_d = cur_d + 1
            s.append({
                "t": "r",
                "v": "\n" + t * cur_d + "}" if p else "}",
                "d": cur_d
            })
            for i in range(len(dpl) - 1, -1, -1):
                dk, dv = dpl[i]
                s.append({
                    "t": "v",
                    "v": dv,
                    "d": child_d
                })
                s.append({
                    "t": "r",
                    "v": f"\"{dk}\": " if p else f"\"{dk}\":",
                    "d": child_d
                })
                if i > 0:
                    s.append({
                        "t": "r",
                        "v": f",\n{t * child_d}" if p else ",",
                        "d": child_d
                    })
            s.append({
                "t": "r",
                "v": f"{{\n{t * child_d}" if p else "{",
                "d": child_d
            })
            continue
        r += f"\"{type(v).__name__}\""
    return r

File: runtimes/ruby/willyhorizont/runtime/xl.rb
module Xl
    def self.escape_string(s)
        return "" if s.nil?
        r = s.to_s
        r = r.gsub("\\", "\\\\")
        r = r.gsub("\"", "\\\"")
        r = r.gsub("\n", "\\n")
        r = r.gsub("\r", "\\r")
        r = r.gsub("\t", "\\t")
        return r
    end
    def self.json_stringify(a, pretty: false)
        p = pretty
        t = " " * 4
        s = [{ "t" => "v", "v" => a, "d" => 0 }]
        r = ""
        while s.length > 0
            c = s.pop
            if c["t"] == "r"
                r += c["v"].to_s
                next
            end
            v = c["v"]
            cur_d = c["d"]
            if v.nil?
                r += "null"
                next
            end
            if v.is_a?(TrueClass) || v.is_a?(FalseClass)
                r += v ? "true" : "false"
                next
            end
            if v.is_a?(String)
                r += "\"" + escape_string(v) + "\""
                next
            end
            if v.is_a?(Numeric)
                r += v.to_s
                next
            end
            if v.is_a?(Proc) || v.is_a?(Method)
                r += "\"[object Function]\""
                next
            end
            if v.is_a?(Array)
                if v.length == 0
                    r += "[]"
                    next
                end
                child_d = cur_d + 1
                s.push({
                    "t" => "r",
                    "v" => p ? "\n" + t * cur_d + "]" : "]",
                    "d" => cur_d
                })
                (v.length - 1).downto(0) do |i|
                    s.push({
                        "t" => "v",
                        "v" => v[i],
                        "d" => child_d
                    })
                    if i > 0
                        s.push({
                            "t" => "r",
                            "v" => p ? ",\n" + t * child_d : ",",
                            "d" => child_d
                        })
                    end
                end
                s.push({
                    "t" => "r",
                    "v" => p ? "[\n" + t * child_d : "[",
                    "d" => child_d
                })
                next
            end
            if v.is_a?(Hash)
                dp_l = v.to_a
                if dp_l.length == 0
                    r += "{}"
                    next
                end
                child_d = cur_d + 1
                s.push({
                    "t" => "r",
                    "v" => p ? "\n" + t * cur_d + "}" : "}",
                    "d" => cur_d
                })
                (dp_l.length - 1).downto(0) do |i|
                    d_k, d_v = dp_l[i]
                    s.push({
                        "t" => "v",
                        "v" => d_v,
                        "d" => child_d
                    })
                    s.push({
                        "t" => "r",
                        "v" => p ? "\"" + d_k.to_s + "\": " : "\"" + d_k.to_s + "\":",
                        "d" => child_d
                    })
                    if i > 0
                        s.push({
                            "t" => "r",
                            "v" => p ? ",\n" + t * child_d : ",",
                            "d" => child_d
                        })
                    end
                end
                s.push({
                    "t" => "r",
                    "v" => p ? "{\n" + t * child_d : "{",
                    "d" => child_d
                })
                next
            end
            r += "\"" + v.class.name + "\""
        end
        return r
    end
end
